unittests/etl/CacheLoaderTest.cpp

```diff
@@ -0,0 +1,141 @@
+        {.key = ripple::uint256{"58CEC9F17733EA7BA68C88E6179B8F207D001EE04D4E0366F958CC04FF6AB834"}, .blob = Blob{'s'}},
+        {.key = ripple::uint256{"64FB1712146BA604C274CC335C5DE7ADFE52D1F8C3E904A9F9765FE8158A3E01"}, .blob = Blob{'s'}},
+        {.key = ripple::uint256{"700BE23B1D9EE3E6BF52543D05843D5345B85D9EDB3D33BBD6B4C3A13C54B38E"}, .blob = Blob{'s'}},
+        {.key = ripple::uint256{"82C297FCBCD634C4424F263D17480AA2F13975DF5846A5BB57246022CEEBE441"}, .blob = Blob{'s'}},
+        {.key = ripple::uint256{"A2AA4C212DC2CA2C49BF58805F7C63363BC981018A01AC9609A7CBAB2A02CEDF"}, .blob = Blob{'s'}},
+        {.key = ripple::uint256{"BC0DAE09C0BFBC4A49AA94B849266588BFD6E1F554B184B5788AC55D6E07EB95"}, .blob = Blob{'s'}},
+        {.key = ripple::uint256{"DCC8759A35CB946511763AA5553A82AA25F20B901C98C9BB74D423BCFAFF5F9D"}, .blob = Blob{'s'}},
+    };
+}
+
+};  // namespace
+
+TEST_F(CacheLoaderTest, FromCache)
+{
+    MockBackend* rawBackendPtr = static_cast<MockBackend*>(mockBackendPtr.get());
+    CacheLoader loader{cfg, ctx, mockBackendPtr, cache};
+
+    auto const diffs = getLatestDiff();
+    ON_CALL(*rawBackendPtr, fetchLedgerDiff(_, _)).WillByDefault(Return(diffs));
+    EXPECT_CALL(*rawBackendPtr, fetchLedgerDiff(_, _)).Times(32);
+
+    auto const loops = diffs.size() + 1;
+    auto const keysSize = 14;
+    std::mutex g_pages_mutex;
+
+    std::map<std::thread::id, uint32_t> threadKeysMap;
+    ON_CALL(*rawBackendPtr, doFetchSuccessorKey(_, SEQ, _))
+        .WillByDefault(Invoke([&]() -> std::optional<ripple::uint256> {
+            // mock the result from doFetchSuccessorKey, be aware this function will be called from multiple threads
+            // for each thread, the last 2 items must be end flag and nullopt, otherwise it will loop forever
+            std::lock_guard<std::mutex> guard(g_pages_mutex);
+            threadKeysMap[std::this_thread::get_id()]++;
+
+            if (threadKeysMap[std::this_thread::get_id()] == keysSize - 1)
+            {
+                return lastKey;
+            }
+            else if (threadKeysMap[std::this_thread::get_id()] == keysSize)
+            {
+                threadKeysMap[std::this_thread::get_id()] = 0;
+                return std::nullopt;
+            }
+            else
+            {
+                return ripple::uint256{INDEX1};
+            }
+        }));
+    EXPECT_CALL(*rawBackendPtr, doFetchSuccessorKey).Times(keysSize * loops);
+
+    ON_CALL(*rawBackendPtr, doFetchLedgerObjects(_, SEQ, _))
+        .WillByDefault(Return(std::vector<Blob>{keysSize - 1, Blob{'s'}}));
+    EXPECT_CALL(*rawBackendPtr, doFetchLedgerObjects).Times(loops);
+
+    EXPECT_CALL(cache, size).Times(AtLeast(1));
+    EXPECT_CALL(cache, update).Times(loops);
+
+    loader.load(SEQ);
+    // Cache loader start a new thread to load the cache, so we need to wait for it to finish
+    auto maxWait = 100;
+    while (maxWait-- != 0 and !cache.isFull())
```

Comment:
This looks like it could be flaky because on top of being slower than it could be since we are relying on polling.
If possible we should rewrite this with some sort of notification/binary_semaphore to minimize the wait time required. I assume that the cache, given the same input (from a mock) will always do the same amount of `doFetchSuccessorKey`s - we can probably use it, wdyt?

Reply:
Good idea. Actually , I think I can mock the setFull, set the condition variable there~ Smart